### PR TITLE
SALTO-2768 align the workflow of PermissionSet to Profile 

### PR DESCRIPTION
**Pull request by Uri-Bar**
_ align the workflow of PermissionSet to Profile _

---
_Additional context for reviewer_:
None

---
_Release Notes_: 
Salesforce-adapter:

* Permission Set - convert some fields from list to map and add change validator
---

_User Notifications_: 
Salesforce-adapter:

* should change the Permission Set type Nacl and the instances as well

### CODE REVIEW

File: packages/salesforce-adapter/src/constants.ts [hunk 1]
```diff
@@ -295,7 +295,7 @@ export const DEFAULT_MAX_INSTANCES_PER_TYPE = 5000
 export const MINIMUM_MAX_ITEMS_IN_RETRIEVE_REQUEST = 500
 export const MAXIMUM_MAX_ITEMS_IN_RETRIEVE_REQUEST = 10000
 export const MAX_QUERY_LENGTH = 2000
-export const DEFAULT_ENUM_FIELD_PERMISSIONS = false
+export const DEFAULT_ENUM_FIELD_PERMISSIONS = true
```
Comment (adamro) [inline packages/salesforce-adapter/src/constants.ts:298]: We are not there yet :)
Leave this to a different PR, we will want to test this properly and we need a migration for existing workspaces to be managed for this little, possibly destructive change to happen

File: packages/salesforce-adapter/src/change_validator.ts [hunk 2]
```diff
@@ -42,7 +42,7 @@ export const changeValidators: Record<ChangeValidatorName, ChangeValidatorCreato
   unknownField: () => unknownFieldValidator,
   customFieldType: () => customFieldTypeValidator,
   standardFieldLabel: () => standardFieldLabelValidator,
-  profileMapKeys: () => profileMapKeysValidator,
+  profilePermissionSetMapKeys: () => profilePermissionSetMapKeysValidator,
```
Comment (adamro) [inline packages/salesforce-adapter/src/change_validator.ts:45]: Ok so to explain what I meant f2f

You could change the above change validator to be a `mapKeysValidator` fn that gets the typeName as a param and just executes the same code as before with the typeName instead of `Profile` where it existed before

File: packages/salesforce-adapter/src/filters/field_permissions_enum.ts [hunk 3]
```diff
@@ -129,7 +159,7 @@ const permissionsEnumToObject = (
 
 const fieldPermissionValuesToEnum = (instance: InstanceElement): InstanceElement => {
   const { fieldPermissions } = instance.value
-  if (fieldPermissions === undefined) {
+  if (fieldPermissions === undefined || _.isArray(fieldPermissions)) {
```
Comment (adamro) [inline packages/salesforce-adapter/src/filters/field_permissions_enum.ts:162]: The warning log doesn't fit this situation
You can split it to 2 if's I guess
Also, why specifically Array? I think the better check here is `!_.isPlainObject`, no?

File: packages/salesforce-adapter/src/filters/field_permissions_enum.ts [hunk 4]
```diff
@@ -166,9 +197,12 @@ const fieldPermissionValuesToObject = (instance: InstanceElement): InstanceEleme
   return instance
 }
 
-const fieldPermissionFieldToEnum = (objectType: ObjectType): void => {
+const fieldPermissionFieldToEnum = async (objectType: ObjectType): Promise<void> => {
   if (isTypeWithFieldPermissions(objectType)) {
-    objectType.fields.fieldPermissions.refType = mapOfMapOfEnumFieldPermissions
+    const fieldType = await objectType.fields.fieldPermissions.getType()
+    if (isMapType(fieldType)) {
```
Comment (adamro) [inline packages/salesforce-adapter/src/filters/field_permissions_enum.ts:203]: I think checking `!isMapType`, logging and returning makes more sense cause we don't expect this to happen and it's a bit too silent like this

File: packages/salesforce-adapter/test/filters/field_permissions_enum.test.ts [hunk 5]
```diff
@@ -18,10 +18,20 @@ import { FilterWith } from '../../src/filter'
 import * as constants from '../../src/constants'
 import fieldPermissionsEnumFilter, { enumFieldPermissions, profileFieldLevelSecurity } from '../../src/filters/field_permissions_enum'
 import { generateProfileType, defaultFilterContext } from '../utils'
+import { METADATA_TYPE, PERMISSION_SET_METADATA_TYPE, SALESFORCE } from '../../src/constants'
 
 describe('FieldPermissionsEnum filter', () => {
```
Comment (adamro) [inline packages/salesforce-adapter/test/filters/field_permissions_enum.test.ts:23]: I think we are missing testing the extra validation logic that you added, right?

**Review by adamro — COMMENTED** [review]
A few comments all around, ping me when you finish an iteration.

Another thought - At the beginning when I wrote the `field_permissions_enum` filter I used Joi to validate everything and it had meaningful (BAD) performance implications. You added a few checks here and some async calls. To not worry about this, I would test our big Salesforce Profiles accounts (NoamEnv & NitzanEnv) to see we didn't create a performance degradation with these extra checks. 
Flows to check - 
1. Fetch
2. Deploy a change to a few fieldPermissions

For Fetch we for sure have the before this change times recorded, for deploy it might be worth checking with a before this change code and then with to compare.

File: packages/salesforce-adapter/src/change_validator.ts (hunk 2 above)
Comment (adamro) [reply packages/salesforce-adapter/src/change_validator.ts:45]: Btw, considering you already wrote the code including the tests etc. and it's not that we do this for similar stuff (ie. the fieldPermissions filter that could be written in a similar way too), you really don't have to bother with this, it's nice to have

File: packages/salesforce-adapter/src/change_validators/profile_permission_set_map_keys.ts [hunk 6]
```diff
@@ -33,20 +47,24 @@ const getMapKeyErrors = async (
   after: InstanceElement
 ): Promise<ChangeError[]> => {
   const errors: ChangeError[] = []
+  const type = await after.getType()
+  const typeName = await apiName(type)
+  const mapper = typeName === PROFILE_METADATA_TYPE
```
Comment (adamro) [inline packages/salesforce-adapter/src/change_validators/profile_permission_set_map_keys.ts:52]: There's a real reason use this and not `metadataTypeToFieldToMapDef`?

Comment (Uri-Bar) [reply packages/salesforce-adapter/src/change_validators/profile_permission_set_map_keys.ts:52]: I will use metadataTypeToFieldToMapDef but I will anyway need the type (as it was used before) and the the typeName for the error detailed message

**Comment by Uri-Bar** [comment]
First fetch measurements: @adamro 
NoamEnv main fetch - 1742.28s user 77.70s system 43% cpu 1:09:16.24 total
Deploy -  217.67s user 8.43s system 103% cpu 3:37.45 total

My branch fetch -  1742.57s user 69.19s system 80% cpu 37:23.52 total
Deploy - 222.46s user 11.31s system 99% cpu 3:54.82 total

Nizanenv main fetch -  441.05s user 21.91s system 87% cpu 8:46.11 total
Deploy -  60.86s user 5.20s system 97% cpu 1:07.47 total

My branch fetch -  395.84s user 18.89s system 81% cpu 8:26.43 total
Deploy -   53.37s user 3.49s system 71% cpu 1:19.68 total

there is no time difference, what else should we check?

File: packages/salesforce-adapter/src/change_validators/profile_permission_set_map_keys.ts [hunk 7]
```diff
@@ -15,16 +15,30 @@
 */
```
Comment (adamro) [inline packages/salesforce-adapter/src/change_validators/profile_permission_set_map_keys.ts:15]: I think we can rename this
`map_keys`
and all the naming around this to the line of `mapKeys`
This is supposed to be a generic change validator that handles validating map keys and not specific for these types, the person who adds the 3rd type doesn't need to add it to the validator name as well

**Comment by adamro** [comment]
Can you change the message of the first commit to have a real meaning and express what it includes? Cause we plan not to squash so this is the actual commit message that will be merged

**Review by adamro — APPROVED** [review]
Notice you have a typo in the commit message + you can drop the details of the first commit, it has no meaning

GJ with this!
